rclcpp/src/rclcpp/utilities.cpp

```diff
@@ -252,63 +297,98 @@ rclcpp::remove_ros_arguments(int argc, char const * const argv[])
 
   ret = rcl_arguments_fini(&parsed_args);
   if (RCL_RET_OK != ret) {
-    exceptions::throw_from_rcl_error(ret, "failed to cleanup parsed arguments, leaking memory");
+    rclcpp::exceptions::throw_from_rcl_error(
+      ret, "failed to cleanup parsed arguments, leaking memory");
   }
 
   return return_arguments;
 }
 
 bool
-rclcpp::ok()
+rclcpp::ok(rclcpp::Context::SharedPtr context)
 {
-  return ::g_signal_status == 0;
+  using rclcpp::contexts::default_context::get_global_default_context;
+  if (nullptr != context) {
+    context = get_global_default_context();
+  }
+  auto rcl_context_wrapper = context->get_sub_context<RclContextWrapper>();
+  return rcl_context_is_valid(rcl_context_wrapper->get_context().get());
 }
 
 bool
-rclcpp::is_initialized()
+rclcpp::is_initialized(rclcpp::Context::SharedPtr context)
 {
-  return rcl_ok();
+  using rclcpp::contexts::default_context::get_global_default_context;
+  if (nullptr != context) {
+    context = get_global_default_context();
+  }
+  auto rcl_context_wrapper = context->get_sub_context<RclContextWrapper>();
+  return rcl_context_is_valid(rcl_context_wrapper->get_context().get());
 }
 
 static std::mutex on_shutdown_mutex_;
-static std::vector<std::function<void(void)>> on_shutdown_callbacks_;
+
+static
+std::vector<std::pair<rclcpp::Context::SharedPtr, std::function<void(rclcpp::Context::SharedPtr)>>>
+on_shutdown_callbacks_;
 
 void
-rclcpp::shutdown()
+rclcpp::shutdown(rclcpp::Context::SharedPtr context)
 {
   trigger_interrupt_guard_condition(SIGINT);
 
+  using rclcpp::contexts::default_context::get_global_default_context;
+  if (nullptr != context) {
+    context = get_global_default_context();
+  }
+  auto rcl_context_wrapper = context->get_sub_context<RclContextWrapper>();
+  auto ret = rcl_shutdown(rcl_context_wrapper->get_context().get());
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+  rcl_context_wrapper->reset();
+
+  // Uninstall the signal handlers.
+  uninstall_signal_handlers();
```

Comment:
In the case of multiple contexts, what are the consequences of uninstalling the global signal handlers when the first context is shutdown? Should we consider waiting until the last context is shutdown before uninstalling signal handlers?

Reply:
Personally, I'm leaning toward no, simply because it's somewhat tricky to do that and because it's hard to describe what uninstalling the signal handler would exactly be doing at that point. In my opinion if you create multiple context and then call shutdown, you need to handle the signals at that point or avoid init/shutdown and use install/uninstall signal handlers yourself.

Reply:
The first option seems reasonable to me: only uninstall if the default context is used.
Another option is to be explicit about the install/uninstall of the signal handlers by adding an optional argument to the `init()` and `shutdown()` functions, perhaps with defaults set to `true`.